### ISSUE TITLE
WebSocket Connection Failed

### ISSUE THREAD

**Issue by techie921** [Microsoft/BotFramework-WebChat#1257]
We are getting below error while clicking on mic button. Any help on this will be appreciated.
WebSocket connection to 'wss://speech.platform.bing.com/speech/recognition/interactive/cognitiveservices/v1?format=simple&language=en-US&Ocp-Apim-Subscription-Key=6f9971474e0041b7aa40d122dc0fb145&X-ConnectionId=AF89B5FFD67145929DD038EDC475FC65' failed: Error during WebSocket handshake: Unexpected response code: 403

**Comment by techie921** [comment]
Hi @compulim,

There is no option to create Bing Speech API Key since it has replaced with new speech service and the webchat is not compatible with newer one. Any help in this case will be appreciated.

**Comment by jawand** [comment]
@compulim the key i am using is perfectly fine. 

if i make a call to this endpoint 

`https://eastus.api.cognitive.microsoft.com/sts/v1.0/issuetoken`
`Ocp-Apim-Subscription-Key: MY_KEY`
`Method: POST`

 i am getting a token from above API call.
and this token is used in below call to Cognitive service.

`WebSocket connection to
wss://speech.platform.bing.com/speech/recognition/interactive/cognitiveservices/v1?format=detailed&language=en-US&Authorization=GENERATED_TOKEN
failed: Error during WebSocket handshake: Unexpected response code: 403`

`ConnectionEstablishErrorEvent | metadata: {} | connectionId: D41CB45AAAAC4092B802115A69BA9886 | statusCode: 1006 | reason: 
OnEvent @ webchat.js:67
OnEvent @ webchat.js:67
OnEvent @ webchat.js:67
i.connectionState.r.ConnectionState.Disconnected.i.connectionEstablishDeferral.i.websocketClient.onclose @ webchat.js:67
`
`RecognitionEndedEvent | metadata: {} | requestId: D303B6BA0EAC4C9D8CA89B20D1EECD51 | audioSourceId: DB3BAA83BF8C4F809C6CA27F5901E80C | audioNodeId: 425C937AB8F6469FBEB2421FD1F5E672 | connectionId: D41CB45AAAAC4092B802115A69BA9886 | authFetchEventId: 09D5A88B0EC942DABF1C8CC931174A26 | status: 7 | error:  | serviceTag: <NULL>
OnEvent @ webchat.js:67
OnEvent @ webchat.js:67
OnEvent @ webchat.js:67
OnComplete @ webchat.js:67
OnConnectionEstablishCompleted @ webchat.js:67
(anonymous) @ webchat.js:67
(anonymous) @ webchat.js:67
ExecuteSuccessCallback @ webchat.js:67
Resolve @ webchat.js:67
Resolve @ webchat.js:67
i.connectionState.r.ConnectionState.Disconnected.i.connectionEstablishDeferral.i.websocketClient.onclose @ webchat.js:67`


Any help is really appreciated.

**Comment by lokeshbansal1993** [comment]
> @compulim the key i am using is perfectly fine.
> 
> if i make a call to this endpoint
> 
> `https://eastus.api.cognitive.microsoft.com/sts/v1.0/issuetoken`
> `Ocp-Apim-Subscription-Key: MY_KEY`
> `Method: POST`
> 
> i am getting a token from above API call.
> and this token is used in below call to Cognitive service.
> 
> `WebSocket connection to wss://speech.platform.bing.com/speech/recognition/interactive/cognitiveservices/v1?format=detailed&language=en-US&Authorization=GENERATED_TOKEN failed: Error during WebSocket handshake: Unexpected response code: 403`
> 
> `ConnectionEstablishErrorEvent | metadata: {} | connectionId: D41CB45AAAAC4092B802115A69BA9886 | statusCode: 1006 | reason: OnEvent @ webchat.js:67 OnEvent @ webchat.js:67 OnEvent @ webchat.js:67 i.connectionState.r.ConnectionState.Disconnected.i.connectionEstablishDeferral.i.websocketClient.onclose @ webchat.js:67 `
> `RecognitionEndedEvent | metadata: {} | requestId: D303B6BA0EAC4C9D8CA89B20D1EECD51 | audioSourceId: DB3BAA83BF8C4F809C6CA27F5901E80C | audioNodeId: 425C937AB8F6469FBEB2421FD1F5E672 | connectionId: D41CB45AAAAC4092B802115A69BA9886 | authFetchEventId: 09D5A88B0EC942DABF1C8CC931174A26 | status: 7 | error: | serviceTag: <NULL> OnEvent @ webchat.js:67 OnEvent @ webchat.js:67 OnEvent @ webchat.js:67 OnComplete @ webchat.js:67 OnConnectionEstablishCompleted @ webchat.js:67 (anonymous) @ webchat.js:67 (anonymous) @ webchat.js:67 ExecuteSuccessCallback @ webchat.js:67 Resolve @ webchat.js:67 Resolve @ webchat.js:67 i.connectionState.r.ConnectionState.Disconnected.i.connectionEstablishDeferral.i.websocketClient.onclose @ webchat.js:67`
> 
> Any help is really appreciated.
> 
> Update:
> 
> * I have tried `websocket.js` from `https://cdn.botframework.com/botframework-webchat/latest/webchat.js` & `https://cdn.botframework.com/botframework-webchat/master/webchat.js`. but none of them works.

what is solutions to ressolved it

**Comment by jawand** [comment]
No, fix as of i know. we changed our approach. we are building a Xamarin forms application and we build our own bot client using DirectLine API's. and now using Cognitive Speech Service REST API's to make Speech to Text and Text to Speech Conversion. But it also has a 15 seconds limit on REST API's. and There is no limit on SDK. But, sadly no SDK for Xamarin.

**Comment by lokeshbansal1993** [comment]
Actually I used in angular but not work...can you suggest any approach for
speech to text

**Comment by jawand** [comment]
I would suggest either wait for it to get fixed.

Or
use REST API's and do it on your own. (Beware there is 15 seconds limit) https://docs.microsoft.com/en-us/azure/cognitive-services/speech-service/rest-apis

Or 
debug the webclient library and check if the library is using old urls. 
as new urls looks like with **region** (eastus, westus, etc.) https://eastus.stt.speech.microsoft.com/speech/recognition/conversation/cognitiveservices/v1?language=en-US

**Comment by jawand** [comment]
**This is a hack.**

ok, i did a quick PoC on changing the URL's from above webclient.js file and it worked.
Here, is what you need to do.

1. download the webclient.js from here https://cdn.botframework.com/botframework-webchat/master/webchat.js

2. Change two endpoints

- find  **https://api.cognitive.microsoft.com/sts/v1.0/issueToken** and replace it with this **https://{your-region}.api.cognitive.microsoft.com/sts/v1.0/issueToken**
- find **wss://speech.platform.bing.com** and replace it with this **wss://{your-region}.stt.speech.microsoft.com**

I haven't changed the URL for Text to Speech but this (**https://speech.platform.bing.com/synthesize**) is what you need to find from the js file. and replace it with new cognitive services URL and see if it works or not.
